tests: improve support for running against minimal targets

dpgeorge opened the pull request:
### Summary

This set of commits tweaks the tests and test runner to allow the full test suite to run against minimal targets.  In particular:
- relevant tests are updated so they skip properly on targets that they don't support
- `run-tests.py` automatically skips more tests when they can't be run on the target (eg if they use floats and the target doesn't enabled float support)
- the unix minimal variant has C stack checking enabled, to help it pass some of the stress tests (and this is anyway a good thing to enable, to protect code from crashing the stack)
- both unix minimal and zephyr minimal targets can now run the "full" test suite without restricting it to certain test directories (instead the test directories are automatically selected like they are for all other targets)
- CI has been updated to run the full test suite on unix minimal and zephyr CI jobs

See the individual commits for more details.

### Testing

Tested locally on unix minimal and zephyr minimal.  Also tested the unix standard test suite to make sure it didn't accidentally skip more tests (it doesn't it still runs the same number of tests).

### Trade-offs and Alternatives

There are lots of small changes here to tests, none of which should change the test behaviour (what it's testing).  All changes are (hopefully!) consistent with existing style for skipping tests.

There are a few changes to the tests to actually allow them to run on minimal targets.  Again, such changes have been carefully made so they don't change what the test is testing.

dpgeorge committed:
tests/stress: Improve qstr_limit test to run on minimal targets.

Changes here are:
- Split out string format module sub-test to a separate test file, so it
  can be skipped on targets that don't have str% capabilities.
- Print the whole type instead of `.__name__` to support targets that don't
  have `.__name__` enabled (this still tests the same thing).
- Print `RuntimeError` exception message to be sure the correct exception
  is being raised.

This test now runs on unix and zephyr minimal configurations.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/misc/sys_exc_info.py: Don't rely on slicing.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Improve skip detection for tests using slice.

This skips some additional tests that use slice, and no longer skips
`basics/string_format_modulo.py` which doesn't actually use slice (but
`float/string_format_modulo.py` does).

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Skip more tests that need slice.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod: Improve skip detection of extmod tests.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod/vfs_posix_paths.py: Use tuple instead of set.

So it can be used on targets without set enabled (or at least not raise a
SyntaxError when compiling it).

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/micropython: Improve skipping of tests using micropython module.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/ci.sh: Enable all possible tests on zephyr CI.

With the recent improvements to the test suite, and fixes for `pow`, the
full test suite can now be run (and appropriate tests will be automatically
skipped).

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/import: Skip import tests where needed.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod/asyncio_heaplock.py: Improve skip detection of test.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Don't include cmdline,io tests for minimal targets.

The unix minimal variant cannot run these tests because:
- it doesn't support -v -v debug printing
- it doesn't have `open()`

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/basics/subclass_native_init.py: Skip if __init__ not supported.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Update list of tests that use float.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/stress/bytecode_limit.py: Tweak to run on minimal builds.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/ports/unix: Improve skip detection for os.getenv and time module.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
unix/variants/minimal: Enable C stack checking.

This unix minimal variant already has a few other things enabled beyond the
default minimal configuration, and C-stack checking is arguably more
important than those, so enable it.  This helps to get some of the stress
tests passing on this variant.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/ci.sh: Enable the full test suite for unix minimal variant.

With all the preceeding improvements to the test suite, it's now possible
to just run `make VARIANT=minimal test` -- which is equivalent to just
`./run-tests.py` -- on the unix minimal variant.

Signed-off-by: Damien George <damien@micropython.org>

projectgus commented on tests/basics/subclass_native_init.py:
```diff
@@ -1,5 +1,9 @@
 # test subclassing a native type and overriding __init__
 
+if not hasattr(object, "__init__"):
```
I wonder if we should make some of these checks explicit on the build type, i.e. `if sys.implementation._build == "minimal"`?

That rules out the possibility of a bug that makes `hasattr(object, "__init__")` return False when it shouldn't. But maybe that's not a very likely bug.

dpgeorge replied:
Unfortunately it won't work for zephyr.

On unix minimal it'll work because we have:
```
>>> sys.implementation
(name='micropython', version=(1, 27, 0, 'preview'), _machine='linux [GCC 15.2.1] version', _build='minimal')
```
but on zephyr minimal we have:
```
>>> sys.implementation
('micropython', (1, 27, 0, 'preview'), 'zephyr-qemu_cortex_m3 with ti_lm3s6965')
```
It doesn't even have attribute names there!  It's just a simple tuple.

We see that `MICROPY_PY_ATTRTUPLE` is enabled explicitly on the unix minimal build, but not on zephyr.  We might actually be able to disable `MICROPY_PY_ATTRTUPLE` on unix minimal with this PR, because we know the tests now run without this enabled (ie because they pass on zephyr minimal).  Probably in the past `MICROPY_PY_ATTRTUPLE` was enabled precisely so it could run some tests.

------

Anyway, TLDR is that it's quite tricky to auto-skip certain tests.  The actual compile-time feature we want to test for here is guarded by `MICROPY_CPYTHON_COMPAT`.  Maybe we can test for that once using a `tests/feature_check` test, and then get `run-tests.py` to have a list of tests to skip that require `MICROPY_CPYTHON_COMPAT` to be enabled?  The downside with that approach is keeping that list in sync with the tests (it'll probably be quite a long list).

projectgus replied:
Thanks for very detailed explanation! I think what's here is fine!

projectgus commented on tools/ci.sh:
```diff
@@ -623,7 +623,10 @@ function ci_unix_minimal_build {
-    (cd tests && MICROPY_CPYTHON3=python3 MICROPY_MICROPYTHON=../ports/unix/build-minimal/micropython ./run-tests.py -e exception_chain -e self_type_check -e subclass_native_init -d basics)
+    # Issues with unix minimal variant tests:
+    # - basics/exception_chain.py output doesn't match because warnings aren't printed
+    # - basics/self_type_check.py crashes because MICROPY_BUILTIN_METHOD_CHECK_SELF_ARG isn't enabled
```
Can we move one or both of these into the test file by checking `sys.implementation`?

projectgus commented on tools/ci.sh:
```diff
@@ -966,7 +969,7 @@ function ci_zephyr_run_tests {
     docker exec zephyr-ci west build -p auto -b qemu_cortex_m3 -- -DCONF_FILE=prj_minimal.conf
     # Issues with zephyr tests:
     # - inf_nan_arith fails pow(-1, nan) test
```
As above, can we move this check into the test?

dpgeorge replied:
I think there's actually a bug in zephyr's libm code, meaning `pow(-1, nan)` doesn't return the correct value.

I don't think our `MICROPY_PY_MATH_POW_FIX_NAN` option fixes that, we would need to add another such workaround.  Probably worth doing.

projectgus reviewed (approved):
I have a few possible suggestions but this looks like a huge step to have test coverage on minimal builds!

dpgeorge committed:
py/modmath: Make MICROPY_PY_MATH_POW_FIX_NAN also fix pow(x, NaN) cases.

This is needed by the zephyr port.  Combining it with the existing
`MICROPY_PY_MATH_POW_FIX_NAN` option should be safe, and eliminates the
need for a separate option.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
zephyr/mpconfigport: Enable MICROPY_PY_MATH_POW_FIX_NAN.

This is needed because zephyr incorrectly has `pow(-1, NaN) = 1`.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/basics: Skip exception_chain and self_type_check on unix minimal.

These two tests can't run on the unix minimal build because it doesn't have
the relevant build options enabled.  So skip them.

Signed-off-by: Damien George <damien@micropython.org>
